src/bin/client.rs

```diff
@@ -133,7 +134,16 @@ impl Client {
             _ => anyhow::bail!("artifact schema is not yet implemented"),
         };
 
-        let req = self
+        let body = if path == Path::new("-") {
+            Body::wrap_stream(ReaderStream::new(GzipEncoder::new(BufReader::new(
+                tokio::io::stdin(),
+            ))))
+        } else {
+            Body::wrap_stream(ReaderStream::new(GzipEncoder::new(BufReader::new(
+                File::open(path).await.context("opening artifact")?,
+            ))))
+        };
```

Comment:
Yeah, I'm not thrilled about that code. Unfortunately, the types need to be the same between the branches:

```
error[E0308]: `if` and `else` have incompatible types
   --> src/bin/client.rs:151:17
    |
148 | /             if path == Path::new("-") {
149 | |                 tokio::io::stdin()
    | |                 ------------------ expected because of this
150 | |             } else {
151 | |                 File::open(path).await.context("opening artifact")?
    | |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Stdin`, found `File`
152 | |             },
    | |_____________- `if` and `else` have incompatible types
```

I could probably box them, but it's probably not worth it.